sound/soc/sof/sof-audio.c

```diff
@@ -107,9 +107,9 @@ int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
 	if (!swidget->private)
 		return 0;
 
-	/* only free when use_count is 0 */
-	if (--swidget->use_count)
-		return 0;
+	/* freeing is only needed at 1->0 */
```

Comment:
@keyonjie in what scenario would you free a widget thats already freed. My guess is that if that happened, you have an imbalance in the setup/free calls and the problem is not here.

Reply:
why not fix the 'wrong logic' then?

Reply:
> why not fix the 'wrong logic' then?

Yes, the wrong logic is fixed together in this PR so this change is only to make it more robust.
On the other hand, @lyakh do you suggest to switch to use kref for this swidget also? Then we will call sof_widget_put() only here, once the refcount of it is decreased to 0, let's call _free() to do the real freeing.

Reply:
> Yes, the wrong logic is fixed together in this PR so this change is only to make it more robust.

I don't get the concept, sorry. Either your logic is robust or it's not.
defensive programming is a different thing, it doesn't make your logic 'more robust' but deals with unexpected failures.